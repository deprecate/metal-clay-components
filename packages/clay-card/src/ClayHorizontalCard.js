import 'clay-checkbox';
import 'clay-dropdown';
import 'clay-link';
import 'clay-sticker';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayHorizontalCard.soy.js';

/**
 * Metal ClayHorizontalCard component.
 */
class ClayHorizontalCard extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayHorizontalCard.STATE = {
	/**
	 * Actions in the dropdown on card.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?array|undefined}
	 * @default undefined
	 */
	actionItems: Config.arrayOf(
		Config.shapeOf({
			label: Config.string().required(),
			separator: Config.bool().value(false),
			url: Config.string(),
		})
	),

	/**
	 * Transform card with checkbox.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?array|undefined}
	 * @default undefined
	 */
	checkbox: Config.shapeOf({
		checked: Config.bool().value(false),
		disabled: Config.bool().value(false),
		indeterminate: Config.bool().value(false),
		name: Config.string(),
		value: Config.string(),
	}),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * Title of the card.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	title: Config.string(),

	/**
	 * Transform title in the url.
	 * @instance
	 * @memberof ClayHorizontalCard
	 * @type {?string}
	 * @default undefined
	 */
	url: Config.string(),
};

defineWebComponent('clay-horizontal-card', ClayHorizontalCard);

Soy.register(ClayHorizontalCard, templates);

export {ClayHorizontalCard};
export default ClayHorizontalCard;
