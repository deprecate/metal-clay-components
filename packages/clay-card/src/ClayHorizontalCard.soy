{namespace ClayHorizontalCard}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param spritemap: string}
	{@param? actionItems: list<[
		label: string,
		separator: bool,
		url: string
	]>}
	{@param? checkbox: [
		checked: bool,
		disabled: bool,
		indeterminate: bool,
		name: string,
		value: string
	]}
	{@param? elementClasses: string}
	{@param? id: string}
	{@param? title: string}
	{@param? url: string}

	{let $attributes kind="attributes"}
		class="card-type-directory
			{if $checkbox}
				{sp}form-check form-check-card form-check-middle-left
			{else}
				{sp}card card-horizontal
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	{let $body kind="html"}
		{call .body}
			{param actionItems: $actionItems /}
			{param spritemap: $spritemap /}
			{param title: $title /}
			{param url: $url /}
		{/call}
	{/let}

	<div {$attributes}>
		{if $checkbox}
			{call ClayCheckbox.render}
				{param checked: $checkbox.checked /}
				{param disabled: $checkbox.disabled /}
				{param indeterminate: $checkbox.indeterminate /}
				{param labelContent kind="html"}
					{let $classesCard kind="text"}
						card card-horizontal
					{/let}

					<div class="{$classesCard}">
						{$body}
					</div>
				{/param}

				{param name: $checkbox.name /}
				{param value: $checkbox.value /}
			{/call}
		{else}
			{$body}
		{/if}
	</div>
{/template}


/**
 * The content the element.
 */
{template .body}
	{@param spritemap: string}
	{@param? actionItems: list<[
		label: string,
		separator: bool,
		url: string
	]>}
	{@param? title: string}
	{@param? url: string}

	<div class="card-body">
		<div class="card-row">
			{let $classesDescription kind="text"}
				flex-col flex-col-expand flex-col-gutters
			{/let}

      {call ClaySticker.render}
        {param icon: 'folder' /}
        {param spritemap: $spritemap /}
        {param style: 'unstyled' /}
      {/call}

			<div class="{$classesDescription}">
				{if $url}
					{call ClayLink.render}
						{param elementClasses: 'card-title text-truncate' /}
						{param href: $url /}
						{param label: $title /}
					{/call}
				{elseif $title}
					<div class="card-title text-truncate" title="{$title}">{$title}</div>
				{/if}
			</div>

			{if $actionItems}
				<div class="flex-col">
					{call ClayActionsDropdown.render}
						{param items: $actionItems /}
						{param spritemap: $spritemap /}
					{/call}
				</div>
			{/if}
		</div>
	</div>
{/template}