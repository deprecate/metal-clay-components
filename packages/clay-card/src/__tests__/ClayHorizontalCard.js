import ClayHorizontalCard from '../ClayHorizontalCard';

let component;
let spritemap = 'icons.svg';
let actionItems = [
	{
		label: 'Edit',
		url: '#1',
	},
	{
		label: 'Save',
		url: '#1',
	},
];

describe('ClayHorizontalCard', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayHorizontalCard({
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with classes', () => {
		component = new ClayHorizontalCard({
			spritemap: spritemap,
			elementClasses: 'my-custom-class',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with id', () => {
		component = new ClayHorizontalCard({
			spritemap: spritemap,
			id: 'myId',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with actionItems', () => {
		component = new ClayHorizontalCard({
			actionItems: actionItems,
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with checkbox checked', () => {
		component = new ClayHorizontalCard({
			checkbox: {
				checked: true,
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with checkbox disabled', () => {
		component = new ClayHorizontalCard({
			checkbox: {
				checked: true,
				disabled: true,
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with checkbox indeterminate', () => {
		component = new ClayHorizontalCard({
			checkbox: {
				indeterminate: true,
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with checkbox `name`', () => {
		component = new ClayHorizontalCard({
			checkbox: {
				checked: true,
				name: 'checkbox01',
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with checkbox `value`', () => {
		component = new ClayHorizontalCard({
			checkbox: {
				checked: true,
				value: 'checkbox',
			},
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with title', () => {
		component = new ClayHorizontalCard({
			title: 'empty-background.png',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayHorizontalCard with url', () => {
		component = new ClayHorizontalCard({
			url: '#1',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});
});
