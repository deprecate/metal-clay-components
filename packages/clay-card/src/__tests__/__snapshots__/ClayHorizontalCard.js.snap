// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayHorizontalCard should render a ClayHorizontalCard with actionItems 1`] = `
<div class="card-type-directory card card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters"></div>
      <div class="flex-col">
        <div class="dropdown dropdown-action">
          <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v">
              <use xlink:href="icons.svg#ellipsis-v"></use>
            </svg>
          </button>
          <div class="dropdown-menu">
            <ul class="list-unstyled">
              <li>
                <a class=" dropdown-item" href="#1">Edit</a>
              </li>
              <li>
                <a class=" dropdown-item" href="#1">Save</a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with checkbox \`name\` 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" name="checkbox01" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with checkbox \`value\` 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" value="checkbox" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with checkbox checked 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with checkbox disabled 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" disabled="disabled" class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with checkbox indeterminate 1`] = `
<div class="card-type-directory form-check form-check-card form-check-middle-left">
  <div class="custom-control custom-checkbox">
    <label>
      <input checked="" class="custom-control-input" ref="input" type="checkbox">
      <span class="custom-control-indicator"></span>
      <div class="card card-horizontal">
        <div class="card-body">
          <div class="card-row">
            <span class="sticker sticker-unstyled">
              <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
                <use xlink:href="icons.svg#folder"></use>
              </svg>
            </span>
            <div class="flex-col flex-col-expand flex-col-gutters"></div>
          </div>
        </div>
      </div>
    </label>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with classes 1`] = `
<div class="card-type-directory card card-horizontal my-custom-class">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters"></div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with id 1`] = `
<div class="card-type-directory card card-horizontal" id="myId">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters"></div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with title 1`] = `
<div class="card-type-directory card card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters">
        <div class="card-title text-truncate" title="empty-background.png">empty-background.png</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render a ClayHorizontalCard with url 1`] = `
<div class="card-type-directory card card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters">
        <a class=" card-title text-truncate" href="#1"></a>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayHorizontalCard should render the default markup 1`] = `
<div class="card-type-directory card card-horizontal">
  <div class="card-body">
    <div class="card-row">
      <span class="sticker sticker-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
          <use xlink:href="icons.svg#folder"></use>
        </svg>
      </span>
      <div class="flex-col flex-col-expand flex-col-gutters"></div>
    </div>
  </div>
</div>
`;
